### DISCUSSION TITLE
Use Case Clarification

### DISCUSSION THREAD

**Discussion by MaheshkumarSundaram** [bpmnServer/bpmn-server#175] [General]
Hi @ralphhanna,

Sometimes, a finished workflow with the same data has to be retriggered from a specific point not from the start due to my business case. I don't think there is any way to restart a finished workflow with the same data from a specific point by skipping some processes in a workflow automatically.

So, I thought to include a 30 day wait time before the actual end event of the workflow. This way if the users need to go back, they would just reroute the workflow to the point from where the processes have to be repeated through a user task.

If I do this 30 day window, there would be minumum 100 workflows at any point of time. Also, my another web application uses `bpmn-client` to query the server simultaneously too (min 50 users would query the server at any point of time). 

Do you see any pitfalls with this approach? Will it affect performance? My concern is scalablitiy of the application.
If you any suggestions to improve, please let me know.

Thanks

**Comment by ralphhanna** [comment]
I would not worry about performance, just it does not look very good on the surface.

Even though bpmn does not support a `restart` option, I am thinking that we can add a new method to restart a completed process at a particular point, it would look like this:

![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/d59c93b1-f096-41c1-a0b5-d70692a6c877)

you can invoke event 'start_2' anytime after the workflow is completed.

If you think this will solve your issue, I can work on this for a next release

Thanks

**Reply by MaheshkumarSundaram** [reply]
Thanks @ralphhanna. Much appreciated.

I suppose this would work as I can put a User Task after Start2, and then users can route it to whatever point they would want to resume it from.

I hope the behaviour I requested is not weird. But I suppose restarting a completed workflow is normal with all the data associated with it.

**Comment by ralphhanna** [comment]
Hi @MaheshkumarSundaram , 

What you are asking for is not weird at all.

Actually, I am adding a new method `engine.restart` to restart an already completed instance at any given point, using same data or different data:

```ts
   restart(instanceQuery:object, startNodeId: string, data:any,user?:ISecureUser, options?:IEngineOptions) :Promise<IExecution>;

```
Should be available later this week.

Thanks

**Reply by MaheshkumarSundaram** [reply]
Thanks @ralphhanna

**Comment by ralphhanna** [comment]
Hi @MaheshkumarSundaram,

This is the default behaviour if you pass empty data {} to restart:
```ts
response=await server.engine.restart({"items.id":itemId},{},user,{});
executed Buy Used Car - completed
item: e0715263-6bf8-4614-b632-1ca674acfffd StartEvent_158yep3 end
item: 0fad3464-9804-4065-a224-3ec91a4a1473 flow_start_buy end
item: de32c872-98d2-4043-ba57-71351c8bbf23 task_Buy end
item: c1af06bd-3b0c-4fac-bc91-d552ae56f39a flow_buy_gw end
item: 2dc4e899-2910-4236-8eba-3e23a5f30aa1 gateway_1 end
item: 1be0587c-f108-4153-9f8a-168a6ec482ae flow_gw1_clean end
item: 8d5b544c-e83e-4918-9e9e-92f29a67288e flow_gw1_repair end
item: 49bc02c8-5838-4a27-b5f7-d1add846bbba task_clean end
item: 9cf21f42-8ff5-4cc0-88bb-f922327cd52e task_repair end
item: 2682cc4d-bb33-4f9d-baaa-af19e7b5ccb1 flow_repair_gw2 end
item: e38a0a1e-a420-41a5-a7e2-59e8a81ebf4f gateway_2 end
item: 17e5d591-04e2-4101-bb9f-7a504f07db0b flow_clean_gw2 end
item: c13c1ad1-ab73-4f3a-be01-45e278607345 gateway_2 end
item: 6540e299-2ed6-4f5a-a53f-80320a5e3ecb flow_gw2_drive end
item: 9506e29f-bd9a-4fd5-9b51-ce35521d39f2 task_Drive end
item: e570afd9-4389-40f3-99c5-73cad4fcede4 flow_drive_end end
item: 128e655f-6bb0-4194-92df-cc90177a16ea Event_19ebav7 end
data: {
  needsCleaning: 'Yes',
  needsRepairs: 'Yes',
  starterUserId: 'user1',
  startDate: 2024-03-05T16:55:27.098Z
}
-------------------- restart -----------------
restarting  de32c872-98d2-4043-ba57-71351c8bbf23
restarted  de32c872-98d2-4043-ba57-71351c8bbf23
-------------------- restart Items-----------------
item: e0715263-6bf8-4614-b632-1ca674acfffd StartEvent_158yep3 end
item: 0fad3464-9804-4065-a224-3ec91a4a1473 flow_start_buy end
item: de32c872-98d2-4043-ba57-71351c8bbf23 task_Buy wait
data: {
  needsCleaning: 'Yes',
  needsRepairs: 'Yes',
  starterUserId: 'user1',
  startDate: 2024-03-05T16:55:27.098Z   
}

```

Any data you pass, will append existing data

**Comment by MaheshkumarSundaram** [comment]
Hi @ralphhanna ,

Thanks for the new release.
 
![308660611-d59c93b1-f096-41c1-a0b5-d70692a6c877](https://github.com/bpmnServer/bpmn-server/assets/55505611/9c03e9fb-e167-4db9-961f-a8ff0aaeac1d)

How would I restart the above workflow from `Start2` but with the complete data from the completed workflow?

**Comment by ralphhanna** [comment]
Sorry, Ignore the diagram above, didn't quite work, so added a new method 'engine.restart`

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna So, you can only restart from a point which was already executed? You can't restart if they weren't in the completed workflow's path?

Is that correct?

**Reply by ralphhanna** [reply]
Correct, in this case just go to a node prior so the user/logic can go the alternate path.

Note that you have to modify the configuration.ts
```ts
		dataStore: function (server) {
			let ds=new DataStore(server);
			ds.enableSavePoints=true;
			return ds;
		},
```
This work by having savePoints part of the instance state, allowing restart to rewind to a previous savePoint.

P.S. The diagram above works if you already know which node that we want to start, by calling `engine.startEvent`

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna, But with `engine.startEvent`, can I restart a completed workflow with the same data without having to provide the data manually? I suppose not.

**Reply by ralphhanna** [reply]
Yes, in both cases, you don't have to provide data again.
### Scenario 1: using a second start event
Before Invoking Start Event:

![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/71807d4b-1087-4902-8ad5-75d63810cd4a)

After

```ts
    response=await api.engine.startEvent(response.id,"start2",{},user,{restart:true});
```
![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/7d4728f5-bd9b-40e9-8ea1-4223399632b8)



- Require a start event not yet started
- If instance is complete, add option restart:true
- Keeps same data
- Repeat processing the flow

### Scenario 2: using restart
Before using restart
![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/732072b8-69fe-4931-adbd-90821a978866)

```ts
 response=await server.engine.restart({"items.id":itemId},{},user,{});
```
After
![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/ed50dfc9-be0f-4f1e-8c33-63f93495cca2)
- Requires a completed instance
- dataStore.enableSavePoints=true;
- need to identify the item to restore to (at start), in the above case 'task_buy'

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna, thank you so much for the examples.
